addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js

```diff
@@ -3996,84 +3996,64 @@ export class OdooEditor extends EventTarget {
-            element[method](p);
-            this.historyStep();
-            if (this._currentStep.selection.anchorNodeOid === 'root') {
-                this._currentStep.selection = {};
-            }
-            return p;
-        }
-        // Handle arrow key presses. Do not insert new 'P'.
+        let nodeBeforeCursor = nodeAfterCursor && nodeAfterCursor.previousElementSibling;
+        // Handle arrow key presses.
         if (currentKeyPress === 'ArrowRight' || currentKeyPress === 'ArrowDown') {
+            // Skip void elements.
+            while (nodeAfterCursor && !nodeAfterCursor.childNodes.length) {
+                nodeAfterCursor = nodeAfterCursor.nextElementSibling;
+            }
             if (nodeAfterCursor) {
                 setSelection(...getDeepestPosition(nodeAfterCursor, 0));
             } else {
-                restoreLastSelection();
+                this.historyResetLatestComputedSelection(true);
             }
         } else if (currentKeyPress === 'ArrowLeft' || currentKeyPress === 'ArrowUp') {
+            // Skip void elements.
+            while (nodeBeforeCursor && !nodeBeforeCursor.childNodes.length) {
+                nodeBeforeCursor = nodeBeforeCursor.previousElementSibling;
+            }
             if (nodeBeforeCursor) {
                 setSelection(...getDeepestPosition(nodeBeforeCursor, nodeSize(nodeBeforeCursor)));
             } else {
-                restoreLastSelection();
+                this.historyResetLatestComputedSelection(true);
             }
-        // Handle other forms of selection. Insert a 'P' if necessary.
-        } else if (!nodeAfterCursor) {
-            // Cursor is at the end of the editable.
-            const p = insertParagraph(this.editable, 'append');
-            setCursorStart(p);
-        } else if (!nodeBeforeCursor) {
-            // Cursor is at the beginning of the editable.
-            const p = insertParagraph(this.editable, 'prepend');
-            setCursorStart(p);
-        } else if (paragraphRelatedElements.includes(nodeAfterCursor.nodeName)) {
-            // Cursor is right before a p.
+        // Handle cursor next to a 'P'.
+        } else if (nodeAfterCursor && paragraphRelatedElements.includes(nodeAfterCursor.nodeName)) {
+            // Cursor is right before a 'P'.
             setCursorStart(nodeAfterCursor);
-        } else if (paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
+        } else if (nodeBeforeCursor && paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
             // Cursor is right after a p.
             setCursorEnd(nodeBeforeCursor);
-        } else {
-            // Cursor is between two non-p blocks
-            const p = insertParagraph(nodeAfterCursor, 'before');
+        // Handle other forms of selection.
+        // Insert a new 'P' is selection resulted from a mouse click.
+        } else if (this._currentMouseState === 'mousedown') {
```

Comment:
@dmo-odoo , this changes a bit the spec that we had agreed upon.
We assumed that any selection change that was not triggered by a key press was necessarily triggered by a mouse click.
But this is not always true, as historyUndo and historyRedo lead to selection changes.
And sometimes undoing a step puts the selection at the editable root. This could lead to an "un-undoable" paragraph insertion. 
Consider the following editable:
```
<hr>
...
```
If the first thing you do when starting editing is placing the cursor before the `hr`, our newly introduced mechanism will insert a `<p>` before it.
But if the user does not want that, (s)he will hit undo. `historyRevert` will remove the `<p>` and, at this point, put the selection on the editable root (at least on my laptop + Chrome). There's no recorded selection on that step, so `historySetSelection` won't change the selection.
As a result the selection is placed before the `<hr>` again, the  mechanism will kick in again and a new `<p>` will be inserted before the `<hr>`. The undo was undone...

So I propose that the paragraph insertion would only happen if the selection change was the result of a mouse click. Otherwise, the selection would simply be removed. Note that this second scenario is rather rare, so it's more like a fallback for weird corner cases.

Reply:
> Isn't that precisely the spec that we agreed upon ?

Yes... we (or just I) just did not think of the scenario where a selection change is done by a history undo.

> I'm not sure I understand. What about changing the selection with the arrows ? Would you remove the selection too ?

No. Arrow key presses are handled like this:
```
 // Handle arrow key presses.
        if (currentKeyPress === 'ArrowRight' || currentKeyPress === 'ArrowDown') {
            // Skip void elements.
            while (nodeAfterCursor && !nodeAfterCursor.childNodes.length) {
                nodeAfterCursor = nodeAfterCursor.nextElementSibling;
            }
            if (nodeAfterCursor) {
                setSelection(...getDeepestPosition(nodeAfterCursor, 0));
            } else {
                this.historyResetLatestComputedSelection(true);
            }
```
The "unsafe" nodes are skipped until a "safe" node is found and the cursor it put there. Unless you reach the end of the document: in this case, the selection if put back where it was. The end result is the cursor not moving (so no "poof"/disappearing).
Theoretically, `this.historyResetLatestComputedSelection(true);` could put the cursor at the editable root, and the function would be triggered again, this time making it "poof"/gone!
But in practice, before the user pressed an arrow, the cursor would be in a "legal" position, so I don't see this scenario happening.

TL;DR: arrow key presses are handled in the 2 first `if` statements. The removal of a selection is really a fallback: it is in the `else` block at the end of the function.

Reply:
What this function does is:
- it puts the cursor on a valid place, being mindful of not undoing an arrow key press movement.
- it inserts an empty paragraph if necessary (no other way to put the cursor between two blocks), but only following a mouse click. No P is inserted by an arrow key press (**1**) or by a history undo (**2**).

We had agreed upon **1** (right?), and now I'm proposing **2**.

Reply:
Ok let's squash those and merge this then, thank you ! 👍